[Mason] Vending step6

myssun0325 opened the pull request:
자판기앱 6단계 구현하였습니다.

주요작업사항
- 구매 기능 추가
- 구매이력관리를 위한 HistoryManager 객체 추가 (+구매이력 복원/저장)
- 구매이력 변경에 대한 Notification 구현
- UIImageView 생성을 위한 ImageViewFactory 객체 추가
- HistoryManager가 구매이력을 userInfo를 통해 데이터를 넘기고 이 데이터를 활용하여 ImageViewFactory에서 이미지뷰를 생성 후 ViewController에서 addSubView를 통한 화면에 구매이력 이미지 추가
- 구매이력 간의 간격 조정을 위해 인덱스 값 (index * 50) 활용

myssun0325 committed:
구매이력관리를 위한 HistoryManager 객체 추가

myssun0325 committed:
VendingMachine의 구매기능 추가

myssun0325 committed:
기존 이미지뷰 레이어 수정

myssun0325 committed:
음료 구매 기능 추가

myssun0325 committed:
UIImageView 생성을 위한 ImageViewFactory객체 추가

myssun0325 committed:
VendingMachine에 구매내역을 불러오는 메서드 추가

myssun0325 committed:
HistoryManager의 Notification post 구현

- 추가로 구매내역을 반환하는 메서드 구현(언아카이빙 시 이 내역을
		바탕으로 구매내역을 그려줌)

myssun0325 committed:
제품을 구매할 때마다 제품이미지 추가

- 가독성을 ViewController의 메서드 순서 정리

myssun0325 committed:
Update README

godrm commented on VendingMachineApp/VendingMachineApp/Model/HistoryManager.swift:
```diff
@@ -0,0 +1,51 @@
+class HistoryManager: NSObject, NSSecureCoding {
+    private var purchased: [Beverage] {
+        didSet {
+            NotificationCenter.default.post(name: .didChangeHistory, object: self, userInfo: ["purchased":purchased])
```
이렇게 하면 배열 자체를 바꾸는게 아니라 객체를 추가할 때마다 계속 불리던가요?

myssun0325 replied:
네. 작은 예제로 테스트해보니 배열의 경우 내부의 값이 변경되면 배열안에 내용이 바뀔때마다 불립니다.
element가 class든지 struct든지 상관없이 불렸습니다.
하지만 만약에 purchased가 배열이 아닌 배열을 갖고 있는 객체라면 그 객체의 배열이 바뀌는것과 상관없이 객체가 변하지 않으면 안불리는 것을 확인했습니다.

godrm replied:
swift array라서 값이 매번 복사되서 새 배열이 만들어져서 그런가 봅니다. 주의할 필요가 있겠습니다 :)

myssun0325 replied:
swift array의 특성 때문에 그렇군요. 저번의 질문하고 확인한 다른 방법으로 추가된 데이터만을 넘기기 위해서 Beverage를 add하는 메서드에서 Notification을 post하는것을 염두해두겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -39,41 +77,47 @@ class ViewController: UIViewController {
+            return
         }
+        let purchasedImageViews = ImageViewFactory.makeImageView(purchased)
+        updatePurchasedImage(purchasedImageViews)
```
* setupPurchasedImage() 함수와 구현 내용이 비슷하지 않나요?
* updatePurchasedImage()를 매번 호출하면 구매할 때마다 전체 구매 이력에 대해 addSubview()가 불려서 1개, 2개, 3개 이런식으로 계속 추가하지 않을까요? View Debugger로 확인해보고 개선해보세요.

myssun0325 replied:
View Debugger로 보고 충격먹었습니다...어마어마한 재구매를 ...ㅎㅎ
왠지 위 피드백사항과 결합하여 수정해야될거같습니다... 개선해보겠습니다.

myssun0325 committed:
피드백반영사항

- 음료이미지들이 Assets에 포함되어 이미지접근 이름 변경
- ImageViewFactory의 이미지뷰 생성 메서드가 이미지뷰배열이 아닌 단일
이미지뷰를 생성
- 구매목록 변경 Notification이 HistoryManager에서 VendingMachine의
메서드에서 post하도록 변경
- ViewController에서 음료수 이미지뷰에 대한 중복코드 리팩토링
- stockImageViews는 뷰자체는 생성되어 있으므로 viewWillAppear에서 뷰의
속성을 변경하도록 수정
- 이미지뷰가 전체구매이력을 addSubView하는 부분수정

myssun0325 commented:
피드백 사항 반영하여 개선하였습니다.

개선사항
- 음료이미지들이 Assets에 포함되어 이미지접근 이름 변경
- ImageViewFactory의 이미지뷰 생성 메서드가 이미지뷰배열이 아닌 단일 이미지뷰를 생성하여 넘김.
- 구매목록 변경 Notification이 HistoryManager에서 VendingMachine의 메서드에서 post하도록 변경
- Notification을 받는 ViewController가 VendingMachine.shared() 객체가 보낸 Notification만을 받도록 수정
- stockImageViews는 뷰자체는 생성되어 있으므로 보여지는 것에 대한 속성변경이라 판단하여 viewWillAppear에서 뷰의 속성을 변경하도록 수정
- 이미지뷰가 전체구매이력을 addSubView하는 부분수정
- 복원된 구매이력과 새로 구매한 것에 대한 구매이력 이미지 추가에 대한 중복코드 수정

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -19,25 +19,31 @@ class ViewController: UIViewController {
-            stockImageViews[$0].layer.masksToBounds = true
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+        stockImageViews.forEach { setupBeverageImageView($0) }
```
viewWillAppear 는 한번만 불리는 메소드 일까요? 여러번 불리는 메소드 일까요?
그러면 forEach로 이미지뷰를 설정하는 메소드는 한 번만 불러도 될까요? 여러번 불러도 될까요?

myssun0325 replied:
😱헛...ㅋㅋㅋ이런....수정하겠습니다

myssun0325 committed:
피드백반영사항 : Stock 이미지뷰 설정을 viewDidLoad에서 하도록 변경

myssun0325 commented:
피드백 주신 부분 반영하였습니다.
viewWillAppear은 화면이 보여질때마다 계속 불리는데..이전과 같이..쓸데 없이 이미지뷰를 계속 설정하게 되네요.
viewDidLoad에서 한번만 설정해주도록 수정하였습니다.
